squidb-processor/src/com/yahoo/squidb/processor/plugins/defaults/UpsertPlugin.java

```diff
@@ -0,0 +1,279 @@
+import java.util.regex.Pattern;
+
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.VariableElement;
+
+/**
+ * A {@link Plugin} that processes the {@link UpsertKey} annotation and generates code to make any table model specs
+ * containing UpsertKey fields implement the Upsertable interface. See that interface for more information. This plugin
+ * can be disabled by passing the {@link PluginEnvironment#OPTIONS_DISABLE_DEFAULT_UPSERT} option in the squidbOptions
+ * key.
+ */
+public class UpsertPlugin extends Plugin {
+
+    private static final DeclaredTypeName UPSERTABLE =
+            new DeclaredTypeName(TypeConstants.SQUIDB_DATA_PACKAGE, "Upsertable");
+
+    private static final DeclaredTypeName CRITERION =
+            new DeclaredTypeName(TypeConstants.SQUIDB_SQL_PACKAGE, "Criterion");
+
+    private static final DeclaredTypeName CRITERION_ARRAY = CRITERION.clone();
+    static {
+        CRITERION_ARRAY.setArrayDepth(1);
+    }
+
+    private static final DeclaredTypeName INDEX =
+            new DeclaredTypeName(TypeConstants.SQUIDB_SQL_PACKAGE, "Index");
+
+    private static final String LOGICAL_KEY_COL_ARRAY_NAME = "logicalKeyColumns";
+
+    private static final Pattern NOT_NULL = Pattern.compile("NOT\\s+NULL", Pattern.CASE_INSENSITIVE);
+
+    private List<TableModelPropertyGenerator> upsertColumns = new ArrayList<>();
+
+    public UpsertPlugin(ModelSpec<?, ?> modelSpec, PluginEnvironment pluginEnv) {
+        super(modelSpec, pluginEnv);
+    }
+
+    @Override
+    public boolean hasChangesForModelSpec() {
+        return modelSpec instanceof TableModelSpecWrapper;
+    }
+
+    @Override
+    public void afterProcessVariableElements() {
+        for (TableModelPropertyGenerator generator : ((TableModelSpecWrapper) modelSpec).getPropertyGenerators()) {
+            VariableElement field = generator.getField();
+            if (field != null && field.getAnnotation(UpsertKey.class) != null) {
+                upsertColumns.add(generator);
+            }
+        }
+        if (!upsertColumns.isEmpty()) {
+            validateUpsertKeyColumns();
+        }
+    }
+
+    private void validateUpsertKeyColumns() {
+        Collections.sort(upsertColumns, new Comparator<PropertyGenerator>() {
+            @Override
+            public int compare(PropertyGenerator o1, PropertyGenerator o2) {
+                int key1Order = o1.getField().getAnnotation(UpsertKey.class).order();
+                int key2Order = o2.getField().getAnnotation(UpsertKey.class).order();
+
+                if (key1Order < key2Order) {
+                    return -1;
+                } else if (key1Order > key2Order) {
+                    return 1;
+                } else {
+                    return 0;
+                }
+            }
+        });
+        checkUpsertColumnOrder();
+        checkUpsertColumnsForNotNullConstraints();
+    }
+
+    // Make sure that the order values were reasonable, i.e. monotonically increasing
```

Comment:
This will be going away in a future enhancement before we release anyways -- just want to get the other code generation refactors in first before I rebuild this part.